src/ngraph/node.cpp

```diff
@@ -46,32 +45,48 @@ Node::Node(const std::string& node_type, const NodeVector& arguments)
             m_inputs.emplace_back(this, i++, output);
         }
     }
+    set_output_size(output_size);
 }
 
-void Node::set_value_type_checked(const element::Type& element_type, const Shape& shape)
+// While we are still doing validation and type inference in the constructor, this is true
+// It can be set to false to debug doing validation/inference after construction. When that
+// is working, these two functions will be removed.
+static bool in_transition = true;
+
+void Node::constructor_validate_and_infer_types()
 {
-    if (m_outputs.size() == 0)
+    if (in_transition)
     {
-        add_output(element_type, shape);
+        validate_and_infer_types();
     }
-    if (element_type != get_element_type() || shape != get_shape())
+}
+
+void Node::delayed_validate_and_infer_types()
+{
+    if (!in_transition)
+    {
+        validate_and_infer_types();
+    }
+}
+
+void Node::set_output_size(size_t n)
+{
+    m_outputs.clear();
+    for (size_t i = m_outputs.size(); i < n; ++i)
     {
-        throw ngraph_error("Setting value type to a different ValueType");
+        auto tensor_view_descriptor = make_shared<descriptor::TensorView>(
+            element::unspecified, Shape(), ngraph::descriptor::Tensor::make_tensor_name(this, i));
```

Comment:
Will the ambiguity between scalar and shape-not-yet-specified be a problem here?

Reply:
We will need some Shape changes. For now we are okay.